File: Simulacion/nPID.m
classdef nPID
    properties
        w
        beta
        y
        P
        Q
        R
        alpha
        v
    end
    
    methods
        function obj = nPID(n, beta, alpha)            
            obj.w = rand(n, 1);
            obj.beta = beta;
            obj.alpha = alpha;
            obj.y = 0;
            obj.P = eye(n) * 10;
            obj.Q = eye(n);
            obj.R = 0.1;
        end
        
        function y = control_u(obj, x, alpha)
            if nargin < 3
                alpha = 0.05;
            end
            
            obj.alpha = alpha;
            obj.v = obj.w' * x;
            obj.y = tanh(obj.v * obj.alpha);
            y = obj.y * obj.beta;
        end
        
        function fit(obj, error, x, eta)
            if nargin < 4
                eta = 0.01;
            end
            
            H = obj.get_H(x);
            H = reshape(H, [3, 1]);
            PH = obj.P * H;
            matriz = obj.R + H' * PH;
            inv_matriz = inv(matriz);
            k = PH * inv_matriz;
            
            delta_w = eta * k * error;
            obj.w = obj.w + delta_w;
            obj.P = obj.P - k * H' * obj.P + obj.Q;
        end
        
        function H = get_H(obj, x)
            disp(x)
            disp(obj.y)
            del_phi = (1 - obj.y * obj.y) * x * obj.beta * obj.alpha;
            H = del_phi;
        end
    end
end

% classdef nPID
%     properties
%         w       % Pesos
%         beta    % Parámetro beta
%         y       % Salida
%         P       % Matriz de ganancia P
%         Q       % Matriz de ganancia Q
%         R       % Matriz de ganancia R
%     end
% 
%     methods
%         function obj = nPID(n, beta)
%             if nargin < 2
%                 n = 3;  % Valor predeterminado para n
%                 beta = 1;  % Valor predeterminado para beta
%             end
%             obj.w = rand(n, 1);
%             obj.beta = beta;
%             obj.y = 0;
%             obj.P = eye(n) * 10;
%             obj.Q = eye(n);
%             obj.R = 0.1;
%         end
% 
%         function u = control_u(obj, x, alpha)
%             if nargin < 3
%                 alpha = 0.05;  % Valor predeterminado para alpha
%             end
%             v = obj.w' * x;
%             obj.y = tanh(v * obj.beta * alpha);
%             u = obj.y * obj.beta;
%         end
% 
%         function fit(obj, error, x, alpha, eta) % Se agrega alpha como argumento
%             if nargin < 5
%                 eta = 0.01;  % Valor predeterminado para eta
%             end
%             H = obj.get_H(x, obj.beta, alpha); % Se pasan beta y alpha como argumentos
%             PH = obj.P * H;
%             matriz = obj.R + H' * PH;
%             invMatriz = pinv(matriz);
%             k = PH * invMatriz;
% 
%             delta_w = eta * k * error;
%             obj.w = obj.w + delta_w;
%             obj.P = obj.P - k * (H' * obj.P) + obj.Q;
%         end
% 
%         function H = get_H(obj, x, beta, alpha) % Se agregan beta y alpha como argumentos
%             H = (1 - obj.y*obj.y) * x * beta * alpha;
%         end
%     end
% end

File: Simulacion/npid.py
import numpy as np

class nPID:
	def __init__(self, n=3, beta=0.3):
		# Inicialización de la clase con parámetros
		self.w = np.random.rand(n, 1)  # Pesos aleatorios
		self.beta = beta  # Parámetro beta
		self.y = 0  # Salida inicial

		# Inicialización de matrices de ganancia P, Q y R
		self.P = np.diag(np.ones(n)) * 10 	#Matriz de estimación del error de los pesos
		self.Q = np.diag(np.ones(n))		#Matriz de estimación del ruido
		self.R = 0.1						#Ruido del error

	def control_u(self, x, alpha=0.05):
		#Método para calcular la salida del controlador
		self.alpha = alpha  # Parámetro alpha
		self.v = np.dot(self.w.T, x)  # Producto punto entre w y x
		self.y = np.tanh(self.v * self.alpha)  # Función tangente hiperbólica
		return self.y * self.beta  # Salida multiplicada por beta

	def fit(self, error, x, eta=0.01):
		# Método para ajustar los pesos del controlador
		H = self.get_H(x)  # Cálculo de la matriz H
		H = H.reshape((3, 1))  # Redimensionamiento de H
		PH = np.dot(self.P, H)  # Producto punto entre P y H
		matriz = self.R + np.dot(H.T, PH)  # Cálculo de matriz
		inv = np.linalg.inv(matriz)  # Inversa de la matriz
		k = np.dot(PH, inv)  # Producto punto entre PH y la inversa

		delta_w = eta * np.dot(k, error)  # Cálculo del cambio en los pesos
		self.w = self.w + delta_w  # Actualización de los pesos
		self.P = self.P - np.dot(k, np.dot(H.T, self.P)) + self.Q  # Actualización de P

	def get_H(self, x):
		# Método para calcular la matriz H
		del_phi = (1 - self.y * self.y) * x * self.beta * self.alpha  # Cálculo de del_phi
		return del_phi  # Retorna H
